issue #359 Simplify programmatic initialization API (#360)
AbsaOSS/spline-spark-agent: README.md

val config = AgentConfig.from(???: Iterable[(String, Any)])

// from a Common Configuration
val config = AgentConfig.from(???: org.apache.commons.configuration.Configuration)

// using a builder
val config = AgentConfig.builder()
  // call some builder methods here...
  .build()

sparkSession.enableLineageTracking(config)
```

**Note**: `AgentConfig` object doesn't override the standard configuration stack. Instead, it serves as an additional configuration mean
with the precedence set between the `spline.properties` and `spline.default.properties` files (see below). 

<a id="configuration"></a>
## Configuration

The agent looks for configuration in the following sources (listed in order of precedence):
- Hadoop configuration (`core-site.xml`)
- Spark configuration
- JVM system properties
- `spline.properties` file on classpath
- `AgentConfig` object
- `spline.default.properties` file on classpath

The file [spline.default.properties](core/src/main/resources/spline.default.properties) contains default values 
for all Spline properties along with additional documentation.
It's a good idea to look in the file to see what properties are available.

> The order of precedence might look counter-intuitive, as one woudl expect that explicitly provided config (`AgentConfig` instance) should 

Reviewer: ```suggestion
The order of precedence might look counter-intuitive, as one would expect that explicitly provided config (`AgentConfig` instance) should 
```

override ones defined in the outer scope. However, prioritizing global config to local one makes it easier to manage Spline settings centrally
on clusters, while still allowing room for customization by job developers.

For example, a company could require lineage metadata from jobs executed on a particular cluster to be sanitized, enhanced with some metrics 

> and credentials and stored in a certain metadata store (a database, file, Spline server etc). The Spline configured needs to be set globally

Reviewer: ```suggestion
and credentials and stored in a certain metadata store (a database, file, Spline server etc). The Spline configuration needs to be set globally
```

and applied to all Spark jobs automatically. However, some jobs might contain hardcoded properties that the developers used locally or on 
a testing environment, and forgot to remove them before submitting jobs into a production.

> In such situation we want cluster settings to have precedence over the hardcoded ones.

Reviewer: Could `hardcoded` be replaced by `local`?

Author: I thought about it, but by _local_ I would understand something that is local to a particular host. Is it correctly to say that, for instance, `System.setProperty(...)` is local settings?

Reviewer: I don't know. What about this:

```suggestion
In such situation we want cluster settings to have precedence over the job settings.
```

> Assuming that hardcoded settings would most likely be defined in thone of the `AgentConfig` object, a property file or a JVM properties, 

Reviewer: thone ?

Author: a typo - removed

on the cluster we could define them in the Spark config or Hadoop config.

> When a property is defined in multiple places, from the perspective of how the conflict is resolved, we can describe two kinds of Spline 
> configuration properties. For majority properties the "first occurrence wins" rule is applied. But there are two that are composed instead

Reviewer: I still think this is sentence could be reworded to make it clearer.
```suggestion
For a properties defined in multiple places Spline has two ways how to resolve the conflict: For majority of properties the "first occurrence wins" rule is applied. But there are two that are composed instead
```

Reviewer: also
`instead of been overridden` -> `instead of being overridden`

> of been overridden: `spline.lineageDispatcher` and `spline.postProcessingFilter`. E.g. if a _LineageDispatcher_ is set to be _Kafka_ in one config

Reviewer: ```suggestion
In case of multiple definitions of property the first occurrence wins, but `spline.lineageDispatcher` and `spline.postProcessingFilter` properties are composed instead. E.g. if a _LineageDispatcher_ is set to be _Kafka_ in one config
```

This sentence seems somewhat complicated to follow. Maybe something like this would work better?
